src/librustdoc/html/render/search_index.rs

```diff
@@ -454,42 +471,323 @@ pub(crate) fn build_index<'tcx>(
+            let mut strbuf = String::new();
+            write_bitmap_to_bytes(&deprecated, &mut buf).unwrap();
+            BASE64_STANDARD.encode_string(&buf, &mut strbuf);
+            crate_data.serialize_field("c", &strbuf)?;
+            strbuf.clear();
+            buf.clear();
+            write_bitmap_to_bytes(&self.emptydesc, &mut buf).unwrap();
+            BASE64_STANDARD.encode_string(&buf, &mut strbuf);
+            crate_data.serialize_field("e", &strbuf)?;
             if has_aliases {
                 crate_data.serialize_field("a", &self.aliases)?;
             }
             crate_data.end()
         }
     }
 
-    // Collect the index into a string
-    format!(
+    let (emptydesc, desc) = {
+        let mut emptydesc = Vec::new();
+        let mut result = Vec::new();
+        let mut set = String::new();
+        let mut len: usize = 0;
+        let mut itemindex: u32 = 0;
+        for desc in std::iter::once(&crate_doc).chain(crate_items.iter().map(|item| &item.desc)) {
+            if desc == "" {
+                emptydesc.push(itemindex);
+                itemindex += 1;
+                continue;
+            }
+            if set.len() >= DESC_INDEX_SHARD_LEN {
+                result.push((len, std::mem::replace(&mut set, String::new())));
+                len = 0;
+            } else if len != 0 {
+                set.push('\n');
+            }
+            set.push_str(&desc);
+            len += 1;
+            itemindex += 1;
+        }
+        result.push((len, std::mem::replace(&mut set, String::new())));
+        (emptydesc, result)
+    };
+
+    let descindex = {
+        let mut descindex = String::with_capacity(desc.len() * 4);
+        for &(len, _) in desc.iter() {
+            write_vlqhex_to_string(len.try_into().unwrap(), &mut descindex);
+        }
+        descindex
+    };
+
+    assert_eq!(
+        crate_items.len() + 1,
+        desc.iter().map(|(len, _)| *len).sum::<usize>() + emptydesc.len()
+    );
+
+    // The index, which is actually used to search, is JSON
+    // It uses `JSON.parse(..)` to actually load, since JSON
+    // parses faster than the full JavaScript syntax.
+    let index = format!(
         r#"["{}",{}]"#,
         krate.name(tcx),
         serde_json::to_string(&CrateData {
-            doc: crate_doc,
             items: crate_items,
             paths: crate_paths,
             aliases: &aliases,
             associated_item_disambiguators: &associated_item_disambiguators,
+            descindex,
+            emptydesc,
         })
         .expect("failed serde conversion")
         // All these `replace` calls are because we have to go through JS string for JSON content.
         .replace('\\', r"\\")
         .replace('\'', r"\'")
         // We need to escape double quotes for the JSON.
         .replace("\\\"", "\\\\\"")
-    )
+    );
+    SerializedSearchIndex { index, desc }
+}
+
+pub(crate) fn write_vlqhex_to_string(n: i32, string: &mut String) {
```

Comment:
Can you move this function and `write_bitmap_to_bytes` into another file please? With a nice doc comment on the module giving extra details about this algorithm.